lib: multi_heap: fix out-of-bounds array access

Nuliu-Z opened the pull request:
Add bounds check to prevent accessing heaps array with index=-1 when the address parameter is NULL or too small.

Return NULL in such cases to match the API specification.

andyross reviewed (commented):
Meh.  No vote.  This is just defensive coding.  `sys_multi_heap_get_heap()` is clearly not intended to check its arguments.  I guess my gut would prefer an __ASSERT (and one that checked both bounds of the selected heap), but this is simple.

Nuliu-Z commented:
@andyross Yeah, Assert is also a valid approach, but but npitre totally read my mind😄.
Based on the sys_multi_heap_free() implementation and the API reference for sys_multi_heap_aligned_alloc, it looks like returning NULL is an expected behavior.
That’s exactly where I go to this — I did a similar check as in sys_multi_heap_free, but it always gives me a pointer back… even though it’s invalid.

```c
 * @param mheap Multi heap pointer
 * @param addr address to be found, must be a pointer to a block allocated by sys_multi_heap_alloc
 * @return 0 multi_heap_rec pointer to a structure to be filled with return data
 *			 or NULL if the heap has not been found
 */
const struct sys_multi_heap_rec *sys_multi_heap_get_heap(const struct sys_multi_heap *mheap,
							 void *addr);
```

```c
void sys_multi_heap_free(struct sys_multi_heap *mheap, void *block)
{
	const struct sys_multi_heap_rec *heap;
	heap = sys_multi_heap_get_heap(mheap, block);
	if (heap != NULL) {
		sys_heap_free(heap->heap, block);
	}
}
```

Nuliu-Z committed:
ib: multi_heap: fix out-of-bounds array access

Add bounds check to prevent accessing heaps array with index=-1
when the address parameter is NULL or too small.

Return NULL in such cases to match the API specification.

Signed-off-by: Nulliu Z <zophaisina@outlook.com>
